crates/fs/public/directory.rs

```diff
@@ -0,0 +1,901 @@
+        path_segments: &[String],
+        store: &B,
+        diverge: bool,
+    ) -> Result<OpResult<Option<Shared<PublicNode>>>> {
+        // Set working node to current directory.
+        let root_node = shared(PublicNode::Dir(self.clone()));
+        let mut working_node = Some(Rc::clone(&root_node));
+
+        // The nodes along the path specified.
+        let mut path_nodes: Vec<(String, Shared<PublicNode>)> = if !path_segments.is_empty() {
+            vec![(String::new(), Rc::clone(&root_node))]
+        } else {
+            vec![]
+        };
+
+        // Iterate over the path segments.
+        for (index, segment) in path_segments.iter().enumerate() {
+            let mut working_node_mut = working_node.as_mut().unwrap().borrow_mut();
+
+            // Fetch node representing the path segment in the working directory.
+            match working_node_mut
+                .as_mut_dir()
+                .lookup_node(segment, store)
+                .await?
+            {
+                Some(found_node) => match &*found_node.borrow() {
+                    // If the node is a directory, set it as the working node.
+                    PublicNode::Dir(_) => {
+                        path_nodes.push((segment.to_string(), Rc::clone(&found_node)));
+
+                        drop(working_node_mut);
+                        working_node = Some(Rc::clone(&found_node));
+                    }
+                    // If the node is a file, we return it if it's the last segment.
+                    PublicNode::File(_) => {
+                        if index != path_segments.len() - 1 {
+                            bail!(FsError::InvalidPath);
+                        }
+
+                        drop(working_node_mut);
+                        working_node = Some(Rc::clone(&found_node));
+                        break;
+                    }
+                },
+                _ => {
+                    // If the node is not found, we return an none.
+                    drop(working_node_mut);
+                    working_node = None;
+                    break;
+                }
+            }
+        }
+
+        // Get nodes which may have diverged.
+        let (root_node, working_node) = if diverge {
+            let diverged_nodes = utils::diverge_and_patch(path_nodes);
+            if !diverged_nodes.is_empty() {
+                (
+                    Rc::clone(&diverged_nodes.first().unwrap().1),
+                    Some(Rc::clone(&diverged_nodes.last().unwrap().1)),
+                )
+            } else {
+                (root_node, working_node)
+            }
+        } else {
+            (root_node, working_node)
+        };
+
+        Ok(OpResult {
+            root_node,
+            result: working_node,
+            diverged: diverge,
+        })
+    }
+
+    /// Looks up a node by its path name in the current directory.
+    ///
+    /// TODO(appcypher): What is a valid path segment identifier?
```

Comment:
Yeah, that's a great question. I can only link an old internal discussion about this: https://talk.fission.codes/t/valid-paths-in-wnfs/2015

My thinking is, let's assume UTF8. Technically any byte array is valid (even ones that utf8-decoded include a slash), because of how the encoding into IPLD works out.

This is something we'd need to think about a little harder if we want to implement the pretty tree maybe.